[SPARK-14922][SPARK-17732][SPARK-23866][SQL] Support partition filter in ALTER TABLE DROP PARTITION and batch dropping PARTITIONS

weixiuli opened the pull request:
### What changes were proposed in this pull request?
Spark only  can drop partitions by exact values. For instance, Spark doesn't support:

>  ALTER TABLE mytable DROP PARTITION(mydate < '2018-04-06') 

The PR adds the support to this syntax.

The PR takes input from the effort in #19691 by @DazhuangSu ,  in #20999 by @mgaido91 ,
, there are some related PRs: #15987,#16036,#19691,#15704.
some related issues:
https://issues.apache.org/jira/browse/SPARK-14922
https://issues.apache.org/jira/browse/SPARK-17732
https://issues.apache.org/jira/browse/SPARK-23866
### How was this patch tested?
Add unittests.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

mgaido91 commented:
please fix all failures

weixiuli commented:
Ping @cloud-fan @mgaido91 @maropu @DazhuangSu   Kindly review, thanks.

maropu commented:
ok to test

SparkQA commented:
**[Test build #112830 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112830/testReport)** for PR 26280 at commit [`ce453ed`](https://github.com/apache/spark/commit/ce453ed06150a4b93136ae9f43e1273a406a6d9e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17765/
Test PASSed.

SparkQA commented:
**[Test build #112830 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112830/testReport)** for PR 26280 at commit [`ce453ed`](https://github.com/apache/spark/commit/ce453ed06150a4b93136ae9f43e1273a406a6d9e).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `public class ExpressionBuilder `
  * `public final class Utilities `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112830/
Test FAILed.

weixiuli commented:
please retry test

weixiuli commented:
jenkins, retest this, please

maropu commented:
retest this please

SparkQA commented:
**[Test build #112845 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112845/testReport)** for PR 26280 at commit [`ce453ed`](https://github.com/apache/spark/commit/ce453ed06150a4b93136ae9f43e1273a406a6d9e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17778/
Test PASSed.

SparkQA commented:
**[Test build #112845 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112845/testReport)** for PR 26280 at commit [`ce453ed`](https://github.com/apache/spark/commit/ce453ed06150a4b93136ae9f43e1273a406a6d9e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `public class ExpressionBuilder `
  * `public final class Utilities `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112845/
Test FAILed.

weixiuli commented:
retest this please

wangyum commented:
retest this please

SparkQA commented:
**[Test build #112919 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112919/testReport)** for PR 26280 at commit [`ce453ed`](https://github.com/apache/spark/commit/ce453ed06150a4b93136ae9f43e1273a406a6d9e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17843/
Test PASSed.

SparkQA commented:
**[Test build #112919 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112919/testReport)** for PR 26280 at commit [`ce453ed`](https://github.com/apache/spark/commit/ce453ed06150a4b93136ae9f43e1273a406a6d9e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `public class ExpressionBuilder `
  * `public final class Utilities `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112919/
Test FAILed.

SparkQA commented:
**[Test build #113000 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113000/testReport)** for PR 26280 at commit [`a0c42fd`](https://github.com/apache/spark/commit/a0c42fd9d2ce107d8e288424a131524df1b0cd2c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17923/
Test PASSed.

SparkQA commented:
**[Test build #113000 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113000/testReport)** for PR 26280 at commit [`a0c42fd`](https://github.com/apache/spark/commit/a0c42fd9d2ce107d8e288424a131524df1b0cd2c).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `public class ExpressionBuilder `
  * `public final class Utilities `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113000/
Test FAILed.

weixiuli commented:
retest this please

maropu commented:
retest this please

SparkQA commented:
**[Test build #113022 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113022/testReport)** for PR 26280 at commit [`a0c42fd`](https://github.com/apache/spark/commit/a0c42fd9d2ce107d8e288424a131524df1b0cd2c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17944/
Test PASSed.

SparkQA commented:
**[Test build #113022 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113022/testReport)** for PR 26280 at commit [`a0c42fd`](https://github.com/apache/spark/commit/a0c42fd9d2ce107d8e288424a131524df1b0cd2c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `public class ExpressionBuilder `
  * `public final class Utilities `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113022/
Test FAILed.

SparkQA commented:
**[Test build #113084 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113084/testReport)** for PR 26280 at commit [`1bd2cd3`](https://github.com/apache/spark/commit/1bd2cd36a291441afb098d73f97abf9e58323b5f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18001/
Test PASSed.

SparkQA commented:
**[Test build #113084 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113084/testReport)** for PR 26280 at commit [`1bd2cd3`](https://github.com/apache/spark/commit/1bd2cd36a291441afb098d73f97abf9e58323b5f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `public class ExpressionBuilder `
  * `public final class Utilities `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113084/
Test FAILed.

SparkQA commented:
**[Test build #113747 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113747/testReport)** for PR 26280 at commit [`e27f13c`](https://github.com/apache/spark/commit/e27f13cdd4f3932d6f33e4811cb1b8facffbe0f5).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18625/
Test PASSed.

SparkQA commented:
**[Test build #113747 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113747/testReport)** for PR 26280 at commit [`e27f13c`](https://github.com/apache/spark/commit/e27f13cdd4f3932d6f33e4811cb1b8facffbe0f5).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `  case class ExpressionBuilder(val table: MetaStoreApiTable,`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113747/
Test FAILed.

weixiuli commented:
retest this please

weixiuli committed:
[SPARK-14922][SPARK-17732][SPARK-23866][SQL] Support partition filters in ALTER TABLE DROP PARTITION and batch dropping PARTITIONS
    - Support partition filters in ALTER TABLE DROP PARTITION.
    - Support batch dropping PARTITIONS by 'ExpressionBuilder'.

SparkQA commented:
**[Test build #113757 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113757/testReport)** for PR 26280 at commit [`21cd780`](https://github.com/apache/spark/commit/21cd7805ef79b9606ae06077f0514e1fefde377c).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18632/
Test FAILed.

SparkQA commented:
**[Test build #113757 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113757/testReport)** for PR 26280 at commit [`21cd780`](https://github.com/apache/spark/commit/21cd7805ef79b9606ae06077f0514e1fefde377c).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `  case class ExpressionBuilder(val table: MetaStoreApiTable,`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113757/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18647/
Test PASSed.

SparkQA commented:
**[Test build #113775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113775/testReport)** for PR 26280 at commit [`0d5bc03`](https://github.com/apache/spark/commit/0d5bc03fd772b1296087ba74169b3012b0acce9b).

SparkQA commented:
**[Test build #113775 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113775/testReport)** for PR 26280 at commit [`0d5bc03`](https://github.com/apache/spark/commit/0d5bc03fd772b1296087ba74169b3012b0acce9b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `  case class ExpressionBuilder(val table: MetaStoreApiTable,`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113775/
Test FAILed.

SparkQA commented:
**[Test build #113834 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113834/testReport)** for PR 26280 at commit [`11b4cb0`](https://github.com/apache/spark/commit/11b4cb0b535528025cb8332cd8b6dd0c1eb17fb3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18706/
Test PASSed.

SparkQA commented:
**[Test build #113834 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113834/testReport)** for PR 26280 at commit [`11b4cb0`](https://github.com/apache/spark/commit/11b4cb0b535528025cb8332cd8b6dd0c1eb17fb3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `  case class ExpressionBuilder(val table: MetaStoreApiTable,`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113834/
Test FAILed.

SparkQA commented:
**[Test build #113852 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113852/testReport)** for PR 26280 at commit [`9d3ede5`](https://github.com/apache/spark/commit/9d3ede515cfa52ebb5223b0965e54c734fa6a422).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18722/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113852/
Test FAILed.

weixiuli commented:
retest this please.
@cloud-fan @mgaido91  @maropu @DazhuangSuKindly review,  thanks.

SparkQA commented:
**[Test build #114062 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114062/testReport)** for PR 26280 at commit [`c2639d0`](https://github.com/apache/spark/commit/c2639d06b57ace7d5eaa7a268434a02c6c1240c1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18922/
Test PASSed.

SparkQA commented:
**[Test build #114062 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114062/testReport)** for PR 26280 at commit [`c2639d0`](https://github.com/apache/spark/commit/c2639d06b57ace7d5eaa7a268434a02c6c1240c1).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PartitioningAttribute(`
  * `  case class ExpressionBuilder(val table: MetaStoreApiTable,`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114062/
Test FAILed.

weixiuli commented:
@cloud-fan Could you review this PR , thanks.

wangshisan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:
```diff
@@ -555,26 +555,114 @@ case class AlterTableDropPartitionCommand(
+      } else {
+        val partitionSpec = filtersSpec.map {
+          case EqualTo(key: Attribute, v: Literal) =>
+            key.name -> v.value.toString
```
Why match EqualTo only? Suppose here would be some LessThan, GreaterThan.. or some unequal to operators, so the partitionSpec would be empty, will it loss such operators?

wangshisan commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -328,6 +328,14 @@ database
+    ;
+
+dropPartitionVal
+    : identifier (comparisonOperator constant)?
```
One suggestion, can we replace constant with expression? In our PROD environment, there are some requirements like 'ALTER TABLE xxx DROP PARTITION (col <  CURRENT_DATE)'.

wangshisan commented:
This is also a requirement in our PROD environment, could anyone help review?
